atc/exec/retry_step.go

```diff
@@ -44,5 +44,8 @@ func (step *RetryStep) Run(ctx context.Context, state RunState) error {
 
 // Succeeded delegates to the last step that it ran.
 func (step *RetryStep) Succeeded() bool {
+	if step.LastAttempt == nil {
```

Comment:
This makes sense and is defensive.
One concern is that since we don't have a way of reproducing this with attempts > 0, this masks a situation where `Succeeded` is being invoked before/concurrently with `Run`, which might be undefined behaviour.

Reply:
I agree, but I am not sure how to avoid this situation. Also, @aoldershaw and I were unable to find a specific case where this happens except if the attempts was set to `0`. so this is to further safeguard against incorrect invocations.

Reply:
also, #5842 should help further safeguard against possible panics like this one.